File: forgerock-auth/src/test/java/org/forgerock/android/auth/DefaultSingleSignOnManagerTest.java
/*
 * Copyright (c) 2019 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth;

import android.content.Context;
import androidx.test.core.app.ApplicationProvider;
import org.forgerock.android.auth.exception.AuthenticationRequiredException;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.io.IOException;
import java.security.GeneralSecurityException;

import static org.junit.Assert.*;

@RunWith(RobolectricTestRunner.class)
public class DefaultSingleSignOnManagerTest {

    private SingleSignOnManager tokenManager;
    private Context context = ApplicationProvider.getApplicationContext();

    @Before
    public void setUp() {
        tokenManager = DefaultSingleSignOnManager.builder()
                .encryptor(new MockEncryptor())
                .context(context).build();
    }

    @After
    public void cleanup() throws Exception {
        tokenManager.clear();
    }

    @Test
    public void storeToken() throws AuthenticationRequiredException {
        SSOToken token = new SSOToken("MyTokenValue");

        tokenManager.persist(token);

        Token storedToken = tokenManager.getToken();
        assertEquals("MyTokenValue", storedToken.getValue());
    }

    @Test
    public void clearToken() throws AuthenticationRequiredException {
        SSOToken token = new SSOToken("MyTokenValue");
        tokenManager.persist(token);
        tokenManager.clear();

        Token storedToken = tokenManager.getToken();
        assertNull(storedToken);
    }

    @Test
    public void getTokenWithError() {
        tokenManager = DefaultSingleSignOnManager.builder()
                .encryptor(new Encryptor() {
                    @Override
                    public byte[] encrypt(byte[] clearText) {
                        return new byte[0];
                    }

                    @Override
                    public byte[] decrypt(byte[] encryptedData) {
                        throw new RuntimeException("Dummy");
                    }

                    @Override
                    public void reset() throws GeneralSecurityException, IOException {

                    }
                })
                .context(context).build();

        Token storedToken = tokenManager.getToken();
        assertNull(storedToken);
    }

    @Test
    public void persistTokenWithRetry() {
        final boolean[] flag = {true};
        tokenManager = DefaultSingleSignOnManager.builder()
                .encryptor(new Encryptor() {
                    @Override
                    public byte[] encrypt(byte[] clearText) {
                        if (flag[0]) {
                            flag[0] = false;
                            throw new RuntimeException();
                        } else {
                            return new byte[0];
                        }
                    }

                    @Override
                    public byte[] decrypt(byte[] encryptedData) {
                        return encryptedData;
                    }

                    @Override
                    public void reset() throws GeneralSecurityException, IOException {

                    }
                })
                .context(context).build();

        SSOToken token = new SSOToken("MyTokenValue");
        tokenManager.persist(token);
        assertNotNull(tokenManager.getToken());

    }
}

File: forgerock-auth/src/test/java/org/forgerock/android/auth/DefaultSingleSignOnManagerTest.kt
/*
 * Copyright (c) 2023 - 2025 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.forgerock.android.auth.storage.Storage
import org.junit.After
import org.junit.Assert
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class DefaultSingleSignOnManagerTest {

    private lateinit var tokenManager: SingleSignOnManager
    private val context = ApplicationProvider.getApplicationContext<Context>()

    private lateinit var ssoTokenStorage: Storage<SSOToken>
    private lateinit var cookiesStorage: Storage<Collection<String>>

    @Before
    fun setUpStorage() {
        ssoTokenStorage = sharedPreferencesStorage<SSOToken>(context = context,
            filename = "ssotoken",
            key = "ssotoken", cacheable = false)
        cookiesStorage = sharedPreferencesStorage<Collection<String>>(context = context,
            filename = "cookies",
            key = "cookies", cacheable = false)
        Options.init(Store(ssoTokenStorage = ssoTokenStorage, cookiesStorage = cookiesStorage))
        tokenManager = DefaultSingleSignOnManager.builder()
            .context(context).build()
    }

    @After
    fun cleanup() {
        ssoTokenStorage.delete()
        cookiesStorage.delete()
        tokenManager.clear()
    }

    @Test
    fun storeToken() {
        val token = SSOToken("MyTokenValue")
        tokenManager.persist(token)
        val storedToken: SSOToken = tokenManager.token
        Assert.assertEquals("MyTokenValue", storedToken.value)
    }

    @Test
    fun clearToken() {
        val token = SSOToken("MyTokenValue")
        tokenManager.persist(token)
        tokenManager.clear()
        val storedToken: SSOToken? = tokenManager.token
        assertNull(storedToken)
    }
}